### PR TITLE
CVSL-619 provide prisoner and com details 

### PR DESCRIPTION
**Pull request by GurnankCheema**
when a previously approved licence is subsequently edited but not re-approved by release date

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt [hunk 1]
```diff
@@ -11,4 +12,26 @@ interface LicenceRepository : JpaRepository<Licence, Long>, JpaSpecificationExec
...
+  fun getLicencesNotApprovedByCRD(): List<UnapprovedLicence>
 }
+
+data class UnapprovedLicence(
+  val crn: String? = null,
+  val forename: String? = null,
+  val surname: String? = null,
+  val comFirstName: String? = null,
+  val comLastName: String? = null,
+  val comEmail: String? = null
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt:36]: Are these all nullable? Would have thought most if not all would be non-nullable

These will need swagger docs too

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt [hunk 2]
```diff
@@ -11,4 +12,26 @@ interface LicenceRepository : JpaRepository<Licence, Long>, JpaSpecificationExec
...
+
+  @Query(
+    """
+    SELECT new uk.gov.justice.digital.hmpps.createandvaryalicenceapi.repository.UnapprovedLicence( l.crn, l.forename, l.surname, com.firstName , com.lastName , com.email)
+        FROM Licence l 
+        JOIN l.submittedBy com
+        WHERE (l.actualReleaseDate = CURRENT_DATE OR l.conditionalReleaseDate = CURRENT_DATE) 
+        AND l.statusCode = 'SUBMITTED'
+        AND EXISTS 
+                (SELECT 1 FROM AuditEvent ae WHERE l.id = ae.licenceId AND ae.detail LIKE '%APPROVED%')
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt:24]: as discussed this is a correlated subquery that could have performance issues with large numbers of matching licences, but think it's a lot more readable than an alternative, it also sounds like its a bit of a stop gap. 

It's also possibly missing indexes

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/LicenceController.kt [hunk 3]
```diff
@@ -1076,4 +1077,41 @@ class LicenceController(private val licenceService: LicenceService) {
...
+      ),
+      ApiResponse(
+        responseCode = "403",
+        description = "Forbidden, requires an appropriate role",
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))],
+      ),
+      ApiResponse(
+        responseCode = "404",
+        description = "Resource not found",
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))],
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/LicenceController.kt:1110]: I think 404 would never happen? you'd just return a 200 and an empty list?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/LicenceController.kt [hunk 4]
```diff
@@ -1076,4 +1077,41 @@ class LicenceController(private val licenceService: LicenceService) {
   ) {
     licenceService.updateSentenceDates(licenceId, request)
   }
+
+  @GetMapping(value = ["/unapproved-by-crd"])
+  @PreAuthorize("hasAnyRole('SYSTEM_USER', 'CVL_ADMIN')")
+  @ResponseBody
+  @Operation(
+    summary = "Get details of edited licences which have not been reapproved by CRD date",
+    description = "Returns prisoner and com details " +
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/LicenceController.kt:1086]: description not quite right?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/LicenceController.kt [hunk 5]
```diff
@@ -1076,4 +1077,41 @@ class LicenceController(private val licenceService: LicenceService) {
   ) {
     licenceService.updateSentenceDates(licenceId, request)
   }
+
+  @GetMapping(value = ["/unapproved-by-crd"])
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/LicenceController.kt:1081]: `not-reapproved-by-crd-date`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt [hunk 6]
```diff
@@ -11,4 +12,26 @@ interface LicenceRepository : JpaRepository<Licence, Long>, JpaSpecificationExec
...
+    """
+  )
+  fun getLicencesNotApprovedByCRD(): List<UnapprovedLicence>
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt:27]: this name doesn't seem to match the logic behind the query?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt [hunk 7]
```diff
@@ -645,7 +650,7 @@ class LicenceService(
       }
 
       val updatedAdditionalConditionUploadSummary = condition.additionalConditionUploadSummary.map {
-        var uploadDetail = additionalConditionUploadDetailRepository.getById(it.uploadDetailId)
+        var uploadDetail = additionalConditionUploadDetailRepository.getReferenceById(it.uploadDetailId)
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt:653]: Is this intentional? it doesn't seem related to your change and there's no associated updated tests?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceIntegrationTest.kt [hunk 8]
```diff
@@ -635,6 +636,51 @@ class LicenceIntegrationTest : IntegrationTestBase() {
...
+    assertThat(result?.get(1)?.surname).isEqualTo("towel")
+    assertThat(result?.get(1)?.comFirstName).isEqualTo("Adam")
+    assertThat(result?.get(1)?.comLastName).isEqualTo("AAA")
+    assertThat(result?.get(1)?.comEmail).isEqualTo("testAAA@probation.gov.uk")
+  }
+
+  // com email address
+// com Name
+// prisonerFirstName prisonerLastName
+// CRN
```
Comment (andrewrlee) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceIntegrationTest.kt:683]: should these comments go?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceIntegrationTest.kt [hunk 9]
```diff
@@ -635,6 +636,51 @@ class LicenceIntegrationTest : IntegrationTestBase() {
...
+  @Sql(
+    "classpath:test_data/clear-all-data.sql"
+  )
+  @Sql(
+    "classpath:test_data/seed-community-offender-manager.sql"
+  )
+  @Sql(
+    "classpath:test_data/seed-audit-events.sql"
+  )
+  @Sql(
```
Comment (andrewrlee) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceIntegrationTest.kt:649]: you can just have one annotation and an array of scripts

File: src/test/resources/test_data/seed-audit-events.sql [hunk 10]
```diff
@@ -3,3 +3,19 @@ values
...
+username,
+full_name,
+event_type,
+summary,
+detail)
+VALUES
+ (4,1,'2022-09-08 10:24:22.000','CVL_OMU','Omu User','USER_EVENT','Licence viewed for Joe Bloggs','ID 1 type AP status APPROVED version 2.0'),
+ (5,2,'2022-09-08 10:24:22.000','CVL_OMU','Omu User','USER_EVENT','Licence viewed for Helen Jones','ID 2 type AP status SUBMITTED version 2.0'),
+ (6,3,'2022-09-08 10:24:22.000','CVL_OMU','Omu User','USER_EVENT','Licence viewed for Jim Smith','ID 3 type AP status APPROVED version 2.0'),
+ (7,2,'2022-09-08 10:24:22.000','CVL_OMU','Omu User','USER_EVENT','Licence viewed for Prison A','ID 4 type AP status SUBMITTED version 2.0');
```
Comment (andrewrlee) [inline src/test/resources/test_data/seed-audit-events.sql:21]: Worth testing all of the conditionals in the query, like having some duplicated approved audit events for the same licence so you can tell you aren't getting duplicate records back

Maybe have someone like "terry towel" who hasn't been submitted? who doesn't appear in the results

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt [hunk 11]
```diff
@@ -136,6 +137,10 @@ class LicenceService(
     return transform(entityLicence)
   }
 
+  fun getLicenceUnapprovedByCrd(): List<UnapprovedLicence> {
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt:140]: pluralise?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt (hunk 7 above)
Comment (mgreen-mojhub) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt:653]: Is this in relation to a deprecated warning?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt (hunk 2 above)
Comment (mgreen-mojhub) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt:24]: This is far from idea but currently we have no other way of knowing if a pre active licence has been previously Approved. There is another ticket to implement versioning on pre Active licences. Once complete, this check will no longer be needed.

The only other method would be to join on the audit table with an inline AND statement. It would require grouping on the licence to avoid duplicates rows brought back by each audit row. 

I don't think performance would be much different.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt (hunk 6 above)
Comment (GurnankCheema) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt:27]: changed to getEditedLicenceNotReApprovedByCrd

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt (hunk 1 above)
Comment (GurnankCheema) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt:36]: correct - made only CRN nullable.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/LicenceController.kt (hunk 3 above)
Comment (GurnankCheema) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/LicenceController.kt:1110]: 404 removed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt (hunk 11 above)
Comment (GurnankCheema) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt:140]: done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt (hunk 7 above)
Comment (GurnankCheema) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt:653]: yes, due to deprecation warning

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceIntegrationTest.kt (hunk 8 above)
Comment (GurnankCheema) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceIntegrationTest.kt:683]: oops!

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceIntegrationTest.kt (hunk 9 above)
Comment (GurnankCheema) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceIntegrationTest.kt:649]: yep

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/LicenceController.kt (hunk 4 above)
Comment (GurnankCheema) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/LicenceController.kt:1086]: moved into its own controller - UnapprovedLicenceController

**Review by mgreen-mojhub — APPROVED** [review]
Looks good
